### PR TITLE
New Apollo Local State API usage (The Return)

### PR DESCRIPTION
**Pull request by estacioneto**
#### What is the purpose of this pull request?

- Use the Apollo new Local State API - [Apollo Client 2.5.0 is out](https://blog.apollographql.com/announcing-apollo-client-2-5-c12230cabbb7)
- The orderForm can be queried from the LinkState. No need to get it from the server.

#### What problem is this solving?

First, we want to delete the orderFormContext (gone too big and messy). To do that, we need to put the orderForm somewhere all our apps can find. This is the Local State API.

#### How should this be manually tested?

[Access the workspace](https://offline--storecomponents.myvtex.com/)

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: react/State.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
+const gql = require('graphql-tag')
```
Comment (lucis) [inline react/State.js]:
```suggestion
import gql from 'graphql-tag'
```

File: react/State.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+const mapToMinicartItem = item => ({
+  seller: null,
+  index: null,
+  parentItemIndex: null,
+  parentAssemblyBinding: null,
+  ...item,
+  __typename: 'MinicartItem',
+})
+
+export const initialState = {
```
Comment (lucis) [inline react/State.js]: I would prefer you to assign initialState and resolvers to variables, and exporting them at the end of the file

File: react/State.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+      })
+      if (orderForm.items) {
+        updateLinkStateItems(orderForm.items, cache)
+      }
+      return orderForm
+    },
+  },
+}
+
+function updateLinkStateItems(newItems, cache) {
```
Comment (lucis) [inline react/State.js]: Any reason to use `function` instead of other function's declarations with `const`?

**Review by lucis — APPROVED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,6 +7,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
 
+### Added
+
+- Use the new Apollo Local State API to manage the orderForm and the minicart items to a faster UX.
+
+## [2.12.4-hkignore] - 2019-03-12
+
+## [2.12.3-hkignore] - 2019-03-12
+
+## [2.12.2-hkignore] - 2019-03-11
```
Comment (lbebber) [inline CHANGELOG.md]: clean this up

File: manifest.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
   "vendor": "vtex",
   "name": "minicart",
   "title": "Mini Cart",
-  "version": "2.12.1",
+  "version": "2.12.4-hkignore",
```
Comment (lbebber) [inline manifest.json]: this probably shouldnt be on this PR either

File: react/__tests__/MiniCart.test.js [hunk 6]
```diff
@@ -1,29 +1,43 @@
...
+import wait from 'waait'
+
 import MiniCart from './../index'
 
 describe('<MiniCart /> component', () => {
+  async function resolveApolloQueries() {
+    // Server queries
+    await wait(0)
+    // Local state queries
+    await wait(0)
```
Comment (lbebber) [inline react/__tests__/MiniCart.test.js:32]: not sure what both these waits do?

**Review by jgfidelis — CHANGES_REQUESTED** [review]
Please test on delivery first :D

File: react/components/MiniCartContent.js [hunk 7]
```diff
@@ -67,43 +71,25 @@ class MiniCartContent extends Component {
...
+    }))
 
     try {
-      await updateAndRefetchOrderForm({
-        variables: {
-          orderFormId: orderForm.orderFormId,
-          items: updatedItem,
-        },
-      })
+      await updateItems(updatedItems)
```
Comment (salesfelipe) [inline react/components/MiniCartContent.js:84]: Shouldn't you add a loading also?

File: react/index.js [hunk 8]
```diff
@@ -213,4 +299,62 @@ miniHOC.schema = {
...
+
+  Component.displayName = `withLinkState(${WrappedComponent.displayName})`
+  Component.propTypes = {
+    client: PropTypes.object.isRequired,
+  }
+  return hoistNonReactStatics(Component, WrappedComponent)
+}
+
+export default compose(
+  graphql(orderForm, { options: () => ({ ssr: false }) }),
```
Comment (salesfelipe) [inline react/index.js:359]:
```suggestion
  graphql(orderForm, { options: { ssr: false } }),
```

File: react/components/MiniCartContent.js [hunk 9]
```diff
@@ -18,34 +20,36 @@ import MiniCartFooter from './MiniCartFooter'
...
     /* Set the mini cart content size */
     isSizeLarge: PropTypes.bool,
     /* Internationalization */
     intl: intlShape.isRequired,
     /** Define a function that is executed when the item is clicked */
-    actionOnClick: PropTypes.func,
+    onClickAction: PropTypes.func,
     /* Reused props */
-    data: MiniCartPropTypes.orderFormContext,
+    orderForm: PropTypes.object,
```
Comment (brunojdo) [inline react/components/MiniCartContent.js:36]: adjust the order for reused props come last

File: react/components/MiniCartContent.js (hunk 7 above)
Comment (estacioneto) [reply react/components/MiniCartContent.js:84]: Why should I? The item is removed from the minicart. This is the background sync magic :sparkles:

**Review by jgfidelis — APPROVED** [review]

**Pull request merged**